Compilers are unable to consume signatures with multi-dimensional arrays of modified types

An attempt to consume the ```Test``` method defined in managed C++ as follows fails:
```
public ref class Test1
{
public:
	virtual void Test(array<long, 2>^ x)
	{
	}
};
```

It can be consumed by VS2013 compilers, but Roslyn gives the following errors:
```
error CS0570: 'Test1.Test(?)' is not supported by the language
error CS0115: 'Test11.Test(int[*,*])': no suitable method found to override
```

Here is the unit-test:
```
        [Fact, WorkItem(4623, "https://github.com/dotnet/roslyn/issues/4623")]
        public void MultiDimensionalArray_01()
        {
            var ilSource = @"
.class public auto ansi beforefieldinit Test1
       extends [mscorlib]System.Object
{
.method public hidebysig specialname rtspecialname 
        instance void  .ctor() cil managed
{
  // Code size       7 (0x7)
  .maxstack  1
  IL_0000:  ldarg.0
  IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
  IL_0006:  ret
} // end of method Test1::.ctor

.method public hidebysig newslot virtual 
        instance void  Test(int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)[0...,0...] x) cil managed
{
  // Code size       11 (0xb)
  .maxstack  1
  IL_0000:  ldstr      ""Test""
  IL_0005:  call       void [mscorlib]System.Console::WriteLine(string)
  IL_000a:  ret
} // end of method Test1::Test
} // end of class Test1
";

            var source = @"
class Test
{
    static void Main()
    {
        Test1 x = new Test1();
        x.Test(null);
        x = new Test11();
        x.Test(null);
    }
}

class Test11 : Test1
{
    public override void Test(int [,] c)
    {
        System.Console.WriteLine(""Overriden"");
    }
}";
            var compilation = CreateCompilationWithCustomILSource(source, ilSource, options: TestOptions.ReleaseExe);

            CompileAndVerify(compilation, expectedOutput: @"Test
Overriden");
        }
```

--- a/src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/ArrayTypeSymbol.cs
@@ -40,7 +40,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
                    rank,
                    declaringAssembly.GetSpecialType(SpecialType.System_Array),
                    GetArrayInterfaces(elementType, rank, declaringAssembly),
-                   customModifiers.NullToEmpty())
+                   customModifiers)
         {
         }
 
@@ -56,13 +56,12 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols
             Debug.Assert(rank >= 1);
             Debug.Assert(constructedInterfaces.Length <= 2);
             Debug.Assert(constructedInterfaces.Length == 0 || rank == 1);
-            Debug.Assert(rank == 1 || !customModifiers.Any());
 
             _elementType = elementType;
             _rank = rank;
             _baseType = array;
             _interfaces = constructedInterfaces;
-            _customModifiers = customModifiers;
+            _customModifiers = customModifiers.NullToEmpty();
         }
 
         private static ImmutableArray<NamedTypeSymbol> GetArrayInterfaces(

--- a/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs
+++ b/src/Compilers/CSharp/Portable/Symbols/Metadata/PE/SymbolFactory.cs
@@ -11,7 +11,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
     {
         internal static readonly SymbolFactory Instance = new SymbolFactory();
 
-        internal override TypeSymbol GetArrayTypeSymbol(PEModuleSymbol moduleSymbol, int rank, TypeSymbol elementType)
+        internal override TypeSymbol GetArrayTypeSymbol(PEModuleSymbol moduleSymbol, int rank, TypeSymbol elementType, ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers)
         {
             if (elementType is UnsupportedMetadataTypeSymbol)
             {
@@ -26,7 +26,7 @@ namespace Microsoft.CodeAnalysis.CSharp.Symbols.Metadata.PE
                 return new UnsupportedMetadataTypeSymbol(); // Found a multi-dimensional array of rank 1 in metadata
             }
 
-            return new ArrayTypeSymbol(moduleSymbol.ContainingAssembly, elementType, ImmutableArray<CustomModifier>.Empty, rank);
+            return new ArrayTypeSymbol(moduleSymbol.ContainingAssembly, elementType, CSharpCustomModifier.Convert(customModifiers), rank);
         }
 
         internal override TypeSymbol GetByRefReturnTypeSymbol(PEModuleSymbol moduleSymbol, TypeSymbol referencedType, ushort countOfCustomModifiersPrecedingByRef)

--- a/src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs
+++ b/src/Compilers/CSharp/Test/Symbol/Symbols/CustomModifiersTests.cs
@@ -1316,5 +1316,59 @@ Test 5
 MyDelegate
 Test 6");
         }
+
+        [Fact, WorkItem(4623, "https://github.com/dotnet/roslyn/issues/4623")]
+        public void MultiDimensionalArray_01()
+        {
+            var ilSource = @"
+.class public auto ansi beforefieldinit Test1
+       extends [mscorlib]System.Object
+{
+    .method public hidebysig specialname rtspecialname 
+            instance void  .ctor() cil managed
+    {
+      // Code size       7 (0x7)
+      .maxstack  1
+      IL_0000:  ldarg.0
+      IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
+      IL_0006:  ret
+    } // end of method Test1::.ctor
+
+    .method public hidebysig newslot virtual 
+            instance void  Test(int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)[0...,0...] x) cil managed
+    {
+      // Code size       11 (0xb)
+      .maxstack  1
+      IL_0000:  ldstr      ""Test""
+      IL_0005:  call       void [mscorlib]System.Console::WriteLine(string)
+      IL_000a:  ret
+    } // end of method Test1::Test
+} // end of class Test1
+";
+
+            var source = @"
+class Test
+{
+    static void Main()
+    {
+        Test1 x = new Test1();
+        x.Test(null);
+        x = new Test11();
+        x.Test(null);
+    }
+}
+
+class Test11 : Test1
+{
+    public override void Test(int [,] c)
+    {
+        System.Console.WriteLine(""Overriden"");
+    }
+}";
+            var compilation = CreateCompilationWithCustomILSource(source, ilSource, options: TestOptions.ReleaseExe);
+
+            CompileAndVerify(compilation, expectedOutput: @"Test
+Overriden");
+        }
     }
 }

--- a/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs
+++ b/src/Compilers/Core/Portable/MetadataReader/MetadataDecoder.cs
@@ -196,7 +196,8 @@ namespace Microsoft.CodeAnalysis
                     int countOfBounds;
                     int countOfLowerBounds;
 
-                    typeSymbol = DecodeTypeOrThrow(ref ppSig, out refersToNoPiaLocalType);
+                    modifiers = DecodeModifiersOrThrow(ref ppSig, out typeCode);
+                    typeSymbol = DecodeTypeOrThrow(ref ppSig, typeCode, out refersToNoPiaLocalType);
                     if (!ppSig.TryReadCompressedInteger(out countOfDimensions) ||
                         !ppSig.TryReadCompressedInteger(out countOfBounds))
                     {
@@ -222,7 +223,7 @@ namespace Microsoft.CodeAnalysis
                         ppSig.TryReadCompressedInteger(out _);
                     }
 
-                    typeSymbol = GetArrayTypeSymbol(countOfDimensions, typeSymbol);
+                    typeSymbol = GetArrayTypeSymbol(countOfDimensions, typeSymbol, modifiers);
                     break;
 
                 case SignatureTypeCode.SZArray:

--- a/src/Compilers/Core/Portable/MetadataReader/SymbolFactory.cs
+++ b/src/Compilers/Core/Portable/MetadataReader/SymbolFactory.cs
@@ -17,7 +17,7 @@ namespace Microsoft.CodeAnalysis
         internal abstract TypeSymbol MakeUnboundIfGeneric(ModuleSymbol moduleSymbol, TypeSymbol type);
 
         internal abstract TypeSymbol GetSZArrayTypeSymbol(ModuleSymbol moduleSymbol, TypeSymbol elementType, ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers);
-        internal abstract TypeSymbol GetArrayTypeSymbol(ModuleSymbol moduleSymbol, int rank, TypeSymbol elementType);
+        internal abstract TypeSymbol GetArrayTypeSymbol(ModuleSymbol moduleSymbol, int rank, TypeSymbol elementType, ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers);
 
         /// <summary>
         /// Produce constructed type symbol.

--- a/src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs
+++ b/src/Compilers/Core/Portable/MetadataReader/TypeNameDecoder.cs
@@ -62,9 +62,9 @@ namespace Microsoft.CodeAnalysis
             return _factory.GetSZArrayTypeSymbol(this.moduleSymbol, elementType, customModifiers);
         }
 
-        protected TypeSymbol GetArrayTypeSymbol(int rank, TypeSymbol elementType)
+        protected TypeSymbol GetArrayTypeSymbol(int rank, TypeSymbol elementType, ImmutableArray<ModifierInfo<TypeSymbol>> customModifiers)
         {
-            return _factory.GetArrayTypeSymbol(this.moduleSymbol, rank, elementType);
+            return _factory.GetArrayTypeSymbol(this.moduleSymbol, rank, elementType, customModifiers);
         }
 
         protected TypeSymbol GetByRefReturnTypeSymbol(TypeSymbol referencedType, ushort countOfCustomModifiersPrecedingByRef)
@@ -209,7 +209,9 @@ namespace Microsoft.CodeAnalysis
                 foreach (int rank in fullName.ArrayRanks)
                 {
                     Debug.Assert(rank > 0);
-                    container = rank == 1 ? GetSZArrayTypeSymbol(container, default(ImmutableArray<ModifierInfo<TypeSymbol>>)) : GetArrayTypeSymbol(rank, container);
+                    container = rank == 1 ? 
+                                GetSZArrayTypeSymbol(container, default(ImmutableArray<ModifierInfo<TypeSymbol>>)) : 
+                                GetArrayTypeSymbol(rank, container, default(ImmutableArray<ModifierInfo<TypeSymbol>>));
                 }
             }
 

--- a/src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb
+++ b/src/Compilers/VisualBasic/Portable/Symbols/ArrayTypeSymbol.vb
@@ -35,7 +35,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
         ''' </summary>
         Friend Sub New(elementType As TypeSymbol, customModifiers As ImmutableArray(Of CustomModifier), rank As Integer, declaringAssembly As AssemblySymbol)
             Me.New(elementType,
-                   customModifiers.NullToEmpty(),
+                   customModifiers,
                    rank,
                    declaringAssembly.GetSpecialType(Microsoft.CodeAnalysis.SpecialType.System_Array),
                    GetArrayInterfaces(elementType, rank, declaringAssembly))
@@ -55,12 +55,11 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols
             Debug.Assert(rank >= 1)
             Debug.Assert(interfaces.Length <= 2)
             Debug.Assert(interfaces.Length = 0 OrElse rank = 1)
-            Debug.Assert(rank = 1 OrElse Not customModifiers.Any())
 
             _elementType = elementType
             _rank = rank
             _systemArray = systemArray
-            _customModifiers = customModifiers
+            _customModifiers = customModifiers.NullToEmpty()
             _interfaces = interfaces
         End Sub
 

--- a/src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/SymbolFactory.vb
+++ b/src/Compilers/VisualBasic/Portable/Symbols/Metadata/PE/SymbolFactory.vb
@@ -10,7 +10,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
 
         Friend Shared ReadOnly Instance As New SymbolFactory()
 
-        Friend Overrides Function GetArrayTypeSymbol(moduleSymbol As PEModuleSymbol, rank As Integer, elementType As TypeSymbol) As TypeSymbol
+        Friend Overrides Function GetArrayTypeSymbol(moduleSymbol As PEModuleSymbol, rank As Integer, elementType As TypeSymbol, customModifiers As ImmutableArray(Of ModifierInfo(Of TypeSymbol))) As TypeSymbol
             If TypeOf elementType Is UnsupportedMetadataTypeSymbol Then
                 Return elementType
             End If
@@ -23,7 +23,7 @@ Namespace Microsoft.CodeAnalysis.VisualBasic.Symbols.Metadata.PE
 
             Return New ArrayTypeSymbol(
                             elementType,
-                            Nothing,
+                            VisualBasicCustomModifier.Convert(customModifiers),
                             rank, moduleSymbol.ContainingAssembly)
         End Function
 

--- a/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/CustomModifiersTests.vb
+++ b/src/Compilers/VisualBasic/Test/Symbol/SymbolsTests/CustomModifiersTests.vb
@@ -1262,5 +1262,61 @@ MyDelegate
 Test 6")
         End Sub
 
+        <Fact, WorkItem(4623, "https://github.com/dotnet/roslyn/issues/4623")>
+        Public Sub MultiDimensionalArray_01()
+            Dim ilSource = <![CDATA[
+.class public auto ansi beforefieldinit Test1
+       extends [mscorlib]System.Object
+{
+    .method public hidebysig specialname rtspecialname 
+            instance void  .ctor() cil managed
+    {
+      // Code size       7 (0x7)
+      .maxstack  1
+      IL_0000:  ldarg.0
+      IL_0001:  call       instance void [mscorlib]System.Object::.ctor()
+      IL_0006:  ret
+    } // end of method Test1::.ctor
+
+    .method public hidebysig newslot virtual 
+            instance void  Test(int32 modopt([mscorlib]System.Runtime.CompilerServices.IsLong)[0...,0...] x) cil managed
+    {
+      // Code size       11 (0xb)
+      .maxstack  1
+      IL_0000:  ldstr      "Test"
+      IL_0005:  call       void [mscorlib]System.Console::WriteLine(string)
+      IL_000a:  ret
+    } // end of method Test1::Test
+} // end of class Test1
+]]>.Value
+            Dim vbSource =
+                <compilation>
+                    <file name="c.vb"><![CDATA[
+Class Module1
+    Shared Sub Main()
+        Dim x As Test1 = new Test1()
+        x.Test(Nothing)
+        x = new Test11()
+        x.Test(Nothing)
+    End Sub
+End Class
+
+class Test11 
+    Inherits Test1
+
+    public overrides Sub Test(c As Integer(,))
+        System.Console.WriteLine("Overriden")
+    End Sub
+End Class
+]]>
+                    </file>
+                </compilation>
+
+            Dim compilation = CreateCompilationWithCustomILSource(vbSource, ilSource, options:=TestOptions.ReleaseExe)
+
+            CompileAndVerify(compilation, expectedOutput:="Test
+Overriden")
+        End Sub
+
     End Class
 End Namespace

